File: apps/worker/worker_python/pipeline/embedding_schema.py
"""Check declared dimensions, including on an empty table; never change DDL."""

from __future__ import annotations

import json
import logging

from worker_python.db import db_connection
from .embedding_contract import (
    EMBEDDING_DIMENSIONS,
    EmbeddingConfig,
    EmbeddingContractError,
    embedding_diagnostic,
    resolve_embedding_config,
)

logger = logging.getLogger(__name__)

EMBEDDING_SCHEMA_SQL = """
    SELECT t.typname AS type_name, a.atttypmod AS dimensions
      FROM pg_attribute a JOIN pg_type t ON t.oid = a.atttypid
     WHERE a.attrelid = to_regclass(%s) AND a.attname = 'embedding'
       AND a.attnum > 0 AND NOT a.attisdropped
"""


def assert_embedding_schema(conn, config: EmbeddingConfig) -> None:
    row = conn.execute(EMBEDDING_SCHEMA_SQL, ("public.scene_embeddings",)).fetchone()
    dimensions = row["dimensions"] if row else None
    if not row or row["type_name"] != "vector" or dimensions != EMBEDDING_DIMENSIONS:
        raise EmbeddingContractError(
            "EMBEDDING_SCHEMA_MISMATCH", "Embedding storage must use vector(1536). Check the database schema.",
            config, db_dimensions=dimensions,
        )
    logger.info(json.dumps(embedding_diagnostic(config, db_dimensions=dimensions)))


def check_embedding_storage() -> None:
    config = resolve_embedding_config()
    with db_connection() as conn:
        assert_embedding_schema(conn, config)

File: apps/api/src/lib/embedding-schema.ts
import {
  EMBEDDING_DIMENSIONS,
  EmbeddingConfigurationError,
  embeddingDiagnostic,
  type EmbeddingConfig,
} from "./embedding-contract";

export const EMBEDDING_SCHEMA_SQL = `
  SELECT t.typname AS type_name, a.atttypmod AS dimensions
    FROM pg_attribute a JOIN pg_type t ON t.oid = a.atttypid
   WHERE a.attrelid = to_regclass($1) AND a.attname = 'embedding'
     AND a.attnum > 0 AND NOT a.attisdropped
`;

type SchemaConnection = {
  query(text: string, values: string[]): Promise<{ rows: { type_name?: unknown; dimensions?: unknown }[] }>;
};

export async function assertEmbeddingSchema(client: SchemaConnection, config: EmbeddingConfig): Promise<void> {
  const { rows } = await client.query(EMBEDDING_SCHEMA_SQL, ["public.scene_embeddings"]);
  const row = rows[0];
  const dimensions = typeof row?.dimensions === "number" ? row.dimensions : null;
  if (row?.type_name !== "vector" || dimensions !== EMBEDDING_DIMENSIONS) {
    throw new EmbeddingConfigurationError(
      "EMBEDDING_SCHEMA_MISMATCH",
      "Embedding storage must use vector(1536). Ask an administrator to check the database schema.",
      config,
      dimensions,
    );
  }
  console.info(JSON.stringify(embeddingDiagnostic(config, { db_dimensions: dimensions })));
}
